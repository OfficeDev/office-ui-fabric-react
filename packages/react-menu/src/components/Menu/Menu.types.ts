import * as React from 'react';
import { ObjectShorthandProps, ShorthandProps } from '@fluentui/react-utilities';
import { MenuListProps } from '../MenuList/index';
import { PositioningProps } from '@fluentui/react-positioning';

/**
 * Extends and drills down Menulist props to simplify API
 * {@docCategory Menu }
 */
export interface MenuProps extends MenuListProps {
  /**
   * Explicitly require children
   */

  children: React.ReactNode;
  /**
   * Whether the popup is open
   */
  open?: boolean;

  /**
   * Whether the popup is open by default
   */
  defaultOpen?: boolean;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup?: ShorthandProps<React.HTMLAttributes<HTMLElement>>;

  /**
   * Where the menu is positioned with respect to the trigger
   */
  position?: PositioningProps['position'];

  /**
   * How the menu is aligned wtih respect to the trigger
   */
  align?: PositioningProps['align'];
}

/**
 * {@docCategory Menu }
 */
export interface MenuState extends MenuProps {
  /**
   * Ref to the root slot
   */
  ref: React.MutableRefObject<HTMLElement>;

  /**
   * Whether the popup is open
   */
  open: boolean;

  /**
   * Callback to open/close the popup
   */
  setOpen: React.Dispatch<React.SetStateAction<boolean>>;

  /**
   * Internal react node that just simplifies handling children
   */
  menuList: React.ReactNode;

  /**
   * Internal react node that just simplifies handling children
   */
  menuTrigger: React.ReactNode;

  /**
   * The ref for the MenuTrigger, used for popup positioning
   */
  triggerRef: React.MutableRefObject<HTMLElement>;

  /**
   * Wrapper to style and add events for the popup
   */
  menuPopup: ObjectShorthandProps<React.HTMLAttributes<HTMLElement>>;
}
