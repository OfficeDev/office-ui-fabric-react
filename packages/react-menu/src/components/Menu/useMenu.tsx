import * as React from 'react';
import { makeMergeProps, resolveShorthandProps, useMergedRefs, useControllableValue } from '@fluentui/react-utilities';
import { usePopper } from '@fluentui/react-positioning';
import { MenuProps, MenuState } from './Menu.types';
import { MenuTrigger } from '../MenuTrigger/index';

export const menuShorthandProps: (keyof MenuProps)[] = ['menuPopup'];

const mergeProps = makeMergeProps<MenuState>({ deepMerge: menuShorthandProps });

/**
 * Create the state required to render Menu.
 *
 * The returned state can be modified with hooks such as useMenuStyles,
 * before being passed to renderMenu.
 *
 * @param props - props from this instance of Menu
 * @param ref - reference to root HTMLElement of Menu
 * @param defaultProps - (optional) default prop values provided by the implementing type
 *
 * {@docCategory Menu }
 */
export const useMenu = (props: MenuProps, ref: React.Ref<HTMLElement>, defaultProps?: MenuProps): MenuState => {
  const state = mergeProps(
    {
      ref: useMergedRefs(ref, React.useRef(null)),
      menuPopup: { as: 'div' },
      position: 'below',
      align: 'start',
    },
    defaultProps,
    resolveShorthandProps(props, menuShorthandProps),
  );

  // TODO Better way to narrow types ?
  const children = React.Children.toArray(state.children) as React.ReactElement[];

  // TODO throw warnings in development safely
  if (children.length !== 2) {
    // eslint-disable-next-line no-console
    console.warn('Menu can only take one MenuTrigger and one MenuList as children');
  }

  const { targetRef: triggerRef, containerRef } = usePopper({ align: state.align, position: state.position });
  state.triggerRef = triggerRef;
  children.forEach(child => {
    if (child.type === MenuTrigger) {
      state.menuTrigger = child;
    } else {
      state.menuList = child;
    }
  });

  state.menuPopup.children = (Component, p) => {
    return (
      // eslint-disable-next-line @typescript-eslint/ban-ts-comment
      // @ts-ignore
      <Component ref={containerRef} {...p}>
        {state.menuList}
      </Component>
    );
  };

  const [open, setOpen] = useControllableValue(state.open, state.defaultOpen);
  // TODO fix useControllableValue typing
  state.open = open !== undefined ? open : state.open;
  state.setOpen = React.useCallback(
    (...args) => {
      setOpen(...args);
    },
    [setOpen],
  );

  return state;
};
