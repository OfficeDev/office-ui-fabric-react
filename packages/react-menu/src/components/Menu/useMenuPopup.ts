import * as React from 'react';
import { getCode, keyboardKey } from '@fluentui/keyboard-key';
import { MenuState } from './Menu.types';
import { isOutsideMenu } from '../../utils/index';

interface UseMenuPopupState
  extends Pick<
    MenuState,
    | 'menuPopup'
    | 'menuPopupRef'
    | 'setOpen'
    | 'menuList'
    | 'onContext'
    | 'onHover'
    | 'triggerId'
    | 'triggerRef'
    | 'open'
  > {}

/**
 * A hook that sets the correct render of the menu popup slot through the children render function
 */
export const useMenuPopup = (state: UseMenuPopupState) => {
  const { menuPopup, menuPopupRef, setOpen, open, menuList, triggerId, triggerRef, onHover, onContext } = state;

  const dismissedWithKeyboardRef = React.useRef(false);
  React.useEffect(() => {
    if (dismissedWithKeyboardRef.current) {
      triggerRef.current?.focus();
      dismissedWithKeyboardRef.current = false;
    }
  }, [triggerRef, dismissedWithKeyboardRef, open]);

  menuPopup.children = (Component, originalProps) => {
    const newProps = { 'aria-labelledby': triggerId, ...originalProps };

    newProps.onMouseEnter = (e: React.MouseEvent<HTMLElement>) => {
      if (onHover && !onContext) {
        setOpen(true);
      }

      originalProps?.onMouseEnter?.(e);
    };

    newProps.onBlur = (e: React.FocusEvent<HTMLElement>) => {
      if (isOutsideMenu({ triggerRef, menuPopupRef, event: e })) {
        setOpen(false);
      }
      originalProps?.onBlur?.(e);
    };

    newProps.onKeyDown = (e: React.KeyboardEvent<HTMLElement>) => {
      const keyCode = getCode(e);
      if (keyCode === keyboardKey.Escape || keyCode === keyboardKey.ArrowLeft) {
        setOpen(false);
        dismissedWithKeyboardRef.current = true;
        e.stopPropagation();
      }

      originalProps?.onKeyDown?.(e);
    };

    newProps.onClick = (e: React.MouseEvent<HTMLElement>) => {
      setOpen(false);
      originalProps?.onClick?.(e);
    };

    return React.createElement(
      Component as React.ElementType,
      {
        ...newProps,
        ref: menuPopupRef,
      },
      menuList,
    );
  };

  return state as MenuState;
};
