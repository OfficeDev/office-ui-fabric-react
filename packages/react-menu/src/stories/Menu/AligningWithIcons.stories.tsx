import * as React from 'react';
import { Menu, MenuTrigger, MenuList, MenuItem, MenuPopover } from '../../index';

import { PasteIcon } from '../../tmp-icons.stories';

export const AligningWithIcons = () => (
  <Menu hasIcons hasCheckmarks>
    <MenuTrigger>
      <button>Toggle menu</button>
    </MenuTrigger>
    <MenuPopover>
      <MenuList>
        <MenuItem>Cut</MenuItem>
        <MenuItem icon={<PasteIcon />}>Paste</MenuItem>
        <MenuItem>Edit</MenuItem>
      </MenuList>
    </MenuPopover>
  </Menu>
);

export default {
  title: 'Components/Menu',
  component: Menu,
};
