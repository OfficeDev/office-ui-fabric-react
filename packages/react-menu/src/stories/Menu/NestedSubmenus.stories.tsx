import * as React from 'react';
import { Menu, MenuTrigger, MenuList, MenuItem, MenuProps, MenuPopover } from '../../index';

const EditorLayoutSubMenu = (props: { controlled?: boolean }) => {
  const { controlled } = props;
  const [open, setOpen] = React.useState(false);
  const onOpenChange: MenuProps['onOpenChange'] = (e, data) => {
    setOpen(data.open);
  };

  const controlledProps = {
    open,
    onOpenChange,
  };

  return (
    <Menu {...(controlled && controlledProps)}>
      <MenuTrigger>
        <MenuItem>Editor Layout</MenuItem>
      </MenuTrigger>

      <MenuPopover>
        <MenuList>
          <MenuItem>Split Up</MenuItem>
          <MenuItem>Split Down</MenuItem>
          <MenuItem>Single</MenuItem>
        </MenuList>
      </MenuPopover>
    </Menu>
  );
};

const AppearanceSubMenu = (props: { controlled?: boolean }) => {
  const { controlled } = props;
  const [open, setOpen] = React.useState(false);
  const onOpenChange: MenuProps['onOpenChange'] = (e, data) => {
    setOpen(data.open);
  };

  const controlledProps = {
    open,
    onOpenChange,
  };

  return (
    <Menu {...(controlled && controlledProps)}>
      <MenuTrigger>
        <MenuItem>Appearance</MenuItem>
      </MenuTrigger>

      <MenuPopover>
        <MenuList>
          <MenuItem>Centered Layout</MenuItem>
          <MenuItem>Zen</MenuItem>
          <MenuItem disabled>Zoom In</MenuItem>
          <MenuItem>Zoom Out</MenuItem>
        </MenuList>
      </MenuPopover>
    </Menu>
  );
};

const PreferencesSubMenu = (props: { controlled?: boolean }) => {
  const { controlled } = props;
  const [open, setOpen] = React.useState(false);
  const onOpenChange: MenuProps['onOpenChange'] = (e, data) => {
    setOpen(data.open);
  };

  const controlledProps = {
    open,
    onOpenChange,
  };

  return (
    <Menu {...(controlled && controlledProps)}>
      <MenuTrigger>
        <MenuItem>Preferences</MenuItem>
      </MenuTrigger>

      <MenuPopover>
        <MenuList>
          <MenuItem>Settings</MenuItem>
          <MenuItem>Online Services Settings</MenuItem>
          <MenuItem>Extensions</MenuItem>
          <AppearanceSubMenu controlled={controlled} />
          <EditorLayoutSubMenu controlled={controlled} />
        </MenuList>
      </MenuPopover>
    </Menu>
  );
};

export const NestedSubmenus = (props: { controlled: boolean }) => {
  const { controlled } = props;

  return (
    <Menu>
      <MenuTrigger>
        <button>Toggle menu</button>
      </MenuTrigger>

      <MenuPopover>
        <MenuList>
          <MenuItem>New </MenuItem>
          <MenuItem>New Window</MenuItem>
          <MenuItem disabled>Open File</MenuItem>
          <MenuItem>Open Folder</MenuItem>
          <PreferencesSubMenu controlled={controlled} />
        </MenuList>
      </MenuPopover>
    </Menu>
  );
};

NestedSubmenus.parameters = {
  layout: 'padded',
};

export const NestedSubmenusControlled = () => <NestedSubmenus controlled />;

NestedSubmenusControlled.parameters = {
  layout: 'padded',
};

export default {
  title: 'Components/Menu',
  component: Menu,
};
