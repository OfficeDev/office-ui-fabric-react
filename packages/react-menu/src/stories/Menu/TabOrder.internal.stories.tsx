import * as React from 'react';
import { Menu, MenuTrigger, MenuList, MenuItem, MenuProps, MenuPopover } from '../../index';

export const TabOrderInternal = (props: Partial<MenuProps>) => (
  <>
    <button>Tabstop</button>
    <Menu {...props}>
      <MenuTrigger>
        <button>Menu trigger</button>
      </MenuTrigger>

      <MenuPopover>
        <MenuList>
          <MenuItem>New </MenuItem>
          <MenuItem>New Window</MenuItem>
          <MenuItem disabled>Open File</MenuItem>
          <MenuItem>Open Folder</MenuItem>
        </MenuList>
      </MenuPopover>
    </Menu>
    <button>Tabstop</button>
  </>
);

export default {
  title: 'Components/Menu',
  component: Menu,
};
