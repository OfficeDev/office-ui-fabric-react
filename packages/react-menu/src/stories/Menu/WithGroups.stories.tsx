import * as React from 'react';
import {
  Menu,
  MenuTrigger,
  MenuList,
  MenuItem,
  MenuGroup,
  MenuDivider,
  MenuGroupHeader,
  MenuPopover,
} from '../../index';

import { CutIcon, PasteIcon, EditIcon } from '../../tmp-icons.stories';

export const WithGroups = () => (
  <Menu>
    <MenuTrigger>
      <button>Toggle menu</button>
    </MenuTrigger>

    <MenuPopover>
      <MenuList>
        <MenuGroup>
          <MenuGroupHeader>Section header</MenuGroupHeader>
          <MenuItem icon={<CutIcon />}>Cut</MenuItem>
          <MenuItem icon={<PasteIcon />}>Paste</MenuItem>
          <MenuItem icon={<EditIcon />}>Edit</MenuItem>
        </MenuGroup>
        <MenuDivider />
        <MenuGroup>
          <MenuGroupHeader>Section header</MenuGroupHeader>
          <MenuItem icon={<CutIcon />}>Cut</MenuItem>
          <MenuItem icon={<PasteIcon />}>Paste</MenuItem>
          <MenuItem icon={<EditIcon />}>Edit</MenuItem>
        </MenuGroup>
      </MenuList>
    </MenuPopover>
  </Menu>
);

export default {
  title: 'Components/Menu',
  component: Menu,
};
