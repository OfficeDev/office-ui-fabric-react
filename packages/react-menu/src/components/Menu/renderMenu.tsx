import * as React from 'react';
import { getSlots, DescendantProvider } from '@fluentui/react-utilities';
import { MenuState } from './Menu.types';
import { menuShorthandProps } from './useMenu';
import { MenuProvider } from '../../contexts/menuContext';
import { MenuDescendantsContext } from '../../contexts/menuDescendantsContext';

/**
 * Render the final JSX of Menu
 * {@docCategory Menu }
 */
export const renderMenu = (state: MenuState) => {
  const { slots, slotProps } = getSlots(state, menuShorthandProps);
  const {
    open,
    setOpen,
    onCheckedValueChange,
    checkedValues,
    defaultCheckedValues,
    onHover,
    onContext,
    triggerRef,
    triggerId,
    menuPopupRef,
  } = state;

  return (
    <MenuProvider
      value={{
        open,
        setOpen,
        onCheckedValueChange,
        checkedValues,
        defaultCheckedValues,
        hasMenuContext: true,
        triggerRef,
        onHover,
        onContext,
        triggerId,
        menuPopupRef,
      }}
    >
      {state.menuTrigger}
      <DescendantProvider
        context={MenuDescendantsContext}
        items={state.triggerDescendants}
        set={state.setTriggerDescendants}
      >
        {state.open && <slots.menuPopup {...slotProps.menuPopup} />}
      </DescendantProvider>
    </MenuProvider>
  );
};
