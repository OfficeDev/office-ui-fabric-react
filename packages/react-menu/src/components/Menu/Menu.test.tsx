import * as React from 'react';
import { Menu } from './Menu';
import * as renderer from 'react-test-renderer';
import { ReactWrapper } from 'enzyme';
import { isConformant } from '../../common/isConformant';

describe('Menu', () => {
  isConformant({
    Component: Menu,
    displayName: 'Menu',
  });

  let wrapper: ReactWrapper | undefined;

  afterEach(() => {
    if (wrapper) {
      wrapper.unmount();
      wrapper = undefined;
    }
  });

  /**
   * Note: see more visual regression tests for Menu in /apps/vr-tests.
   */
  it('renders a default state', () => {
    const component = renderer.create(<Menu>Default Menu</Menu>);
    const tree = component.toJSON();
    expect(tree).toMatchSnapshot();
  });
});
