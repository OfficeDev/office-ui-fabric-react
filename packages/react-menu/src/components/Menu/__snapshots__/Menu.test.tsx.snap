// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Menu renders a default state 1`] = `
<button
  aria-haspopup={true}
  id="1"
  onClick={[Function]}
>
  Menu trigger
</button>
`;
